barter/examples/custom_strategy.rs

```diff
@@ -0,0 +1,414 @@
+        // Update strategy state when we receive a specific MarketEvent
+    }
+}
+
+impl AlgoStrategy for BuyAndHoldStrategy {
+    type State = EngineState;
+
+    fn generate_algo_orders(
+        &self,
+        state: &Self::State,
+    ) -> (
+        // TODO: Use default types for orders ExchangeKey and InstrumentKey?
+        impl IntoIterator<Item = Order<ExchangeIndex, InstrumentIndex, RequestCancel>>,
+        impl IntoIterator<Item = Order<ExchangeIndex, InstrumentIndex, RequestOpen>>,
+    ) {
+        dbg!(state.strategy.desired_quantity_to_buy);
+
+        // We already bought as much as we wanted
+        if state.strategy.desired_quantity_to_buy.is_zero() {
+            return (vec![], vec![]);
+        }
+
+        // There are already some opened orders. Skip the current evaluation.
+        // This is here so that we don't push many buy orders for the same
+        // desired quantity, before even receiving an event of execution.
+        // TODO: How should this atomicity actually be handled? There are cases
+        // when we can have orders as marked completed and no trade events yet
+        // received from the exchange.
+        let mut current_orders = state
+            .instruments
+            .instrument_index(&self.instrument)
+            .orders
+            .orders();
+        if current_orders.any(|o| o.state.is_open_or_in_flight()) {
+            return (vec![], vec![]);
+        }
+
+        // Instrument data we are trading
+        let instrument = state.instruments.instrument_index(&self.instrument);
+
+        // Current market price. This is needed because of how the MockExchange
+        // currently mocks order executions. Ideally we wouldn't need to pass
+        // this in the future.
+        // TODO: Ugly type annotation needed.
+        let Some(price) =
+            <DefaultMarketData as MarketDataState<InstrumentIndex>>::price(&instrument.market)
+        else {
+            warn!("Market price is not yet set");
+            return (vec![], vec![]);
+        };
+
+        // Available quote balance for the instrument
+        let available_quote = state
+            .assets
+            .asset_index(&instrument.instrument.underlying.quote)
+            .balance
+            .expect("balance should exist for the quote asset")
+            .value
+            .free;
+
+        info!(%available_quote, "Available quote amount");
+
+        // Buy order we are proposing
+        let order = Order {
+            exchange: instrument.instrument.exchange,
+            instrument: self.instrument,
+            strategy: self.id.clone(),
+            cid: ClientOrderId::default(),
+            side: Side::Buy,
+            state: RequestOpen {
+                kind: OrderKind::Market,
+                time_in_force: TimeInForce::ImmediateOrCancel,
+                price,
+                quantity: state.strategy.desired_quantity_to_buy,
+            },
+        };
+
+        (vec![], vec![order])
```

Comment:
Since you have multiple return types you'd need to do some itertools::Either magic on them

Reply:
Or you could use (Option, Option), since this impls IntoIterator